[ACQ-554] Style filter dropdowns on /curriculum

levadadenys opened the pull request:
### [ACQ-554] Style filter dropdowns on /curriculum

* styled filter dropdowns on /curriculum
* some refactor
* some cleanup


Before:
<img width="1086" alt="Знімок екрана 2023-05-26 о 17 23 05" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/cead5f3c-f7f7-4066-805b-6fcbea1fad99">
<img width="1086" alt="Знімок екрана 2023-05-26 о 17 23 11" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/ad2529bb-ea7d-472b-bf7b-99d9938697c3">

After:
<img width="1086" alt="Знімок екрана 2023-05-26 о 17 41 02" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/6f6bc833-ab7c-4bfb-8b36-35c1fe6ba2a4">
<img width="1086" alt="Знімок екрана 2023-05-26 о 17 41 09" src="https://github.com/code-dot-org/code-dot-org/assets/22244040/815077f6-1be8-462d-be3e-17957711f551">


<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->

## Links
- jira ticket: [ACQ-554](https://codedotorg.atlassian.net/browse/ACQ-554?atlOrigin=eyJpIjoiMTI3ZTRhMzg2MmJmNGNmMDhmMjdjOTQxYjIyMTE1NjIiLCJwIjoiaiJ9)

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

levadadenys commented:
Button changes will be moved to https://github.com/code-dot-org/code-dot-org/pull/52117 once that PR is merged

megcrenshaw commented on apps/src/templates/CheckboxDropdown.jsx:
```diff
@@ -12,63 +18,89 @@ const CheckboxDropdown = ({
+            title={i18n.filterCheckIconTitle({filter_label: label})}
+          />
+        )}
+        <Typography semanticTag="span" visualAppearance="body-one">
```
Really cool to see the Typography component coming to life with these! It's very clear what's going on and makes styling a breeze!

megcrenshaw replied:
Thanks for your thoughtfulness with this.

megcrenshaw commented on apps/src/templates/CheckboxDropdown.jsx:
```diff
@@ -12,63 +18,89 @@ const CheckboxDropdown = ({
+              className={style.affectAllButton}
+              text={i18n.selectAll()}
+              onClick={onSelectAll}
+              styleAsText
```
TIL about this prop!

megcrenshaw commented on apps/src/templates/checkbox-dropdown.module.scss:
```diff
@@ -1,10 +1,100 @@
+    //    border: 2px solid blue;
+    //    border-radius: 4px;
+    //  }
+    //}
```
Nit: could you delete this before merging?

levadadenys replied:
yep, can be deleted already (checkbox is ready in different PR)

megcrenshaw commented on apps/style/code-studio/curriculum_catalog_container.module.scss:
```diff
@@ -2,11 +2,18 @@
 }
 
-.catalogClearFiltersButton {
+button.catalogClearFiltersButton {
```
I'm not seeing where we use the `button` tag anymore. Are these styles being applied?

megcrenshaw replied:
Oh I see –– the Button component uses the `button` tag. Ignore me!

megcrenshaw commented on apps/src/templates/curriculumCatalog/CurriculumCatalog.jsx:
```diff
@@ -290,14 +294,14 @@ const CurriculumCatalog = ({curriculaData, isEnglish}) => {
+        <Button
           id="clear-filters"
-          type="button"
           className={style.catalogClearFiltersButton}
```
Same question here –– not convinced these styles are doing anything at this point

megcrenshaw reviewed (commented):
This is looking great, and I'm loving the coding patterns! I am worried that the `preventDefault()` methods are cascading into some unintended keyboard behavior. In particular, pressing "enter" when on a checkbox makes *all* the checkboxes selected, we aren't able to hit "enter" to trigger "clear filters" (but the "space" key does clear filters), and the "esc" key opens the dropdown.

It's a bit hard to understand what's happening in the video because you can't see the keys I'm pressing, but here's the video:
https://github.com/code-dot-org/code-dot-org/assets/9142121/9c03a49f-0ef9-442a-9cbb-9fc2923673d1

levadadenys commented:
> This is looking great, and I'm loving the coding patterns! I am worried that the `preventDefault()` methods are cascading into some unintended keyboard behavior. In particular, pressing "enter" when on a checkbox makes _all_ the checkboxes selected, we aren't able to hit "enter" to trigger "clear filters" (but the "space" key does clear filters), and the "esc" key opens the dropdown.
> 
> It's a bit hard to understand what's happening in the video because you can't see the keys I'm pressing, but here's the video: https://github.com/code-dot-org/code-dot-org/assets/9142121/9c03a49f-0ef9-442a-9cbb-9fc2923673d1

Yeah, but it's not because of preventDefault. The reason for space/enter issue is that our buttons didn't have the type prop on it, and so they had submit type by default and that's how button type submit behaves in currents structure.
Since button updates from https://github.com/code-dot-org/code-dot-org/pull/52117 are merged, I've fixed the issue with enter/space bar. Now the checkbox will be updated once you click space (default input checkbox behavior) and all the buttons can be clicked via space or enter.

The escape logic for main dropdown button - not sure if it's a bug or default behavior, but I suggest we look deeper on it in separate PR since the idea of this PR is to give basic styling updates and not to block @code-dot-org/acquisition-products-team in case something won't from design system won't be ready on time.

levadadenys commented:
^^^ fixed issues from comments in last commit

megcrenshaw commented:
The space and enter keys are working as expected on the dropdowns! The one exception is the "Clear filters" button –– that should work with the "enter" key but it's only working with the "space" key (which is different behavior from what's in production).

Agreed that the weird escape behavior is a leftover bug from before, and no need to address that in your styling work. @TurnerRiley is it on your radar that the "escape" key is opening dropdowns? It should be that the "escape" keys only closes dropdowns.

megcrenshaw reviewed (approved):
Merging this as-is introduces a bug not present earlier: that the "Clear filters" button doesn't get triggered on pressing "enter." It does sound like this PR is blocking other work, so I'm going to approve it.

Two tickets are added to our backlog:
1) https://codedotorg.atlassian.net/browse/ACQ-637 (escape key on dropdowns)
2) https://codedotorg.atlassian.net/browse/ACQ-636 (enter key on clear filters button)
It's unclear to me who owns #2, but #1 will definitely be tackled by someone on our team!

This looks great –– thanks for your thoughtful work on this.

breville commented:
Nice to see these improvements.  Some quick thoughts from playing with it:

- While each row has a grey hover highlight if you are anywhere over the row, it's only clickable when over the text or checkbox part of the row.  It would be good to make the entire row clickable, and for the cursor to reflect that.
- When toggling an option, it's easy to do a double-click, leaving some dropdown text highlighted.  It would be good to disable selection of the text.
- When any option is selected in a dropdown, a checkmark appears at the top, shifting the following dropdowns.  It might be good to create space for that checkmark at all times so that the buttons don't shift around.
- It might be desirable for touch target sizing, but each row seems quite big, and for the longer lists of dropdowns, some users will have to scroll their window up and down to see all options.  It might be worth shrinking everything a bit.
- A smaller detail, but the baseline of the "Filter by" text doesn't match the baseline of the dropdown texts.
- Should the down caret used here match the down caret used elsewhere, for example in the header?

breville commented:
Some additional notes:

- We don't appear to be using HTML `select` tags for dropdowns.  This means we don't get the convenience of browser-provided dropdowns, and on mobile devices in particular we don't get their special touch-friendly implementations.  
  - Do we prefer some benefits from a custom implementation?
  - Are there accessibility impacts because of this?
  - See https://code.org/learn for examples of using `select` tags for dropdowns.
- On narrower browser widths for mobile devices, the buttons get a bit messy
  - For example, the checkmark is positioned inconsistently and the buttons end up different sizes.

<img width="683" alt="Screenshot 2023-06-09 at 8 17 45 AM" src="https://github.com/code-dot-org/code-dot-org/assets/2205926/4e5d310b-c0be-4f42-aa0a-aa57f0af5f41">

dmcavoy commented:
The down caret is probably a design team question. Meg can you ask in the #design channel? 

For mobile we are not prioritizing making the mobile perfect right now just doable.

megcrenshaw commented:
Asked in the design channel here: https://codedotorg.slack.com/archives/C0SUN2SSF/p1686330140352379

One more thing I noticed: the focus state on the dropdowns don't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well?

levadadenys commented:
Wow guys, the pr was merged a while ago and I've just seen all the comments appeared here after merge. Let me try answer all the questions one by one:
@breville 
* Fixed in the second PR - `While each row has a grey hover highlight if you are anywhere over the row, it's only clickable when over the text or checkbox part of the row. It would be good to make the entire row clickable, and for the cursor to reflect that.`
* Nice catch, can be fixed with couple lines of css, thank you! - `When toggling an option, it's easy to do a double-click, leaving some dropdown text highlighted. It would be good to disable selection of the text.`
* Nice catch, but should be discussed with the designers - `When any option is selected in a dropdown, a checkmark appears at the top, shifting the following dropdowns. It might be good to create space for that checkmark at all times so that the buttons don't shift around.`
* I believe it's also fixed or partially fixed into second PR. Changing more then that should be discussed with designers - `It might be desirable for touch target sizing, but each row seems quite big, and for the longer lists of dropdowns, some users will have to scroll their window up and down to see all options. It might be worth shrinking everything a bit.`
* Nice catch, should be fixed with 1 line of css. `A smaller detail, but the baseline of the "Filter by" text doesn't match the baseline of the dropdown texts.`
* This dropdown is part of new Design System (DSCO) so the answer is no, we have other icons for new dropdown - `Should the down caret used here match the down caret used elsewhere, for example in the header?`

levadadenys commented:
@breville 
* Oh, that's a great question. Yeah, I guess select will be a better solution if we can implement all the custom styles we need from the designs - `We don't appear to be using HTML select tags for dropdowns. This means we don't get the convenience of browser-provided dropdowns, and on mobile devices in particular we don't get their special touch-friendly implementations.`

* Nice catch as well, need to discuss it with design team `On narrower browser widths for mobile devices, the buttons get a bit messy`

levadadenys commented:
@megcrenshaw 
* Nice catch, thank you! That's again as with checkbox my mistake (honestly don't know how it happened). Will fix it. - `One more thing I noticed: the focus state on the dropdowns doesn't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well?`

levadadenys commented:
@breville @megcrenshaw @dmcavoy 
Thanks a lot for your attention and notes! Sorry for long response. Hope I answered all of your questions now.
Please, ping me in slack if it takes me more then 1 day to respond in github or you're commenting merged PR.

levadadenys commented:
@megcrenshaw also, to double check and clarify:
the issue when filter by section has more width then the screensize should be done by @code-dot-org/acquisition-products-team once designs for that are ready, cause that's part of page functionality, not design system (unless it'll be part of it by that time/will use some page/section template from DSCO).
Making those dropdown buttons look as expected is however my responsibility, so I'll check on how to fix the checkmark is positioned inconsistently and the buttons end up different sizes. (fixed in [pr](https://github.com/code-dot-org/code-dot-org/pull/52318))

megcrenshaw commented:
That sounds good to me, thanks for the clarifying, Denys!

TurnerRiley commented:
> Asked in the design channel here: https://codedotorg.slack.com/archives/C0SUN2SSF/p1686330140352379
> 
> One more thing I noticed: the focus state on the dropdowns doesn't have high enough contrast for accessibility. That's a design thing –– @TurnerRiley can you include that in the Jira ticket(s) you're writing as well? <img alt="image" width="308" src="https://user-images.githubusercontent.com/9142121/244759913-e2111cac-c223-4935-9ecc-2cff6e3a2524.png">

On it! Thank you @megcrenshaw and @levadadenys for the work on this!
